src/qcrypto/simbasics.py

```diff
@@ -0,0 +1,80 @@
+import numpy as np
+import dataclasses
+
+@dataclasses.dataclass
+class Qubit:
+    """
+    Data class which represents the qubit and its quantum state.
+    Quantum state is given by:
+    |psi> = cos(theta/2) |0> + e^{i*phi} * sin(theta/2) |1>
+    """
+
+    def __init__(self, theta, phi, base):
+        self.base = base
+        self.state = np.array([
+            np.cos(theta/2),
+            np.exp(1j * phi) * np.sin(theta/2)
+        ])
+
+    def __str__(self):
+        return "{} |0> + {} |1>".format(self.state[0], self.state[1])
+    
+    def __repr__(self):
+        return "{} |0> + {} |1>".format(self.state[0], self.state[1])
+
+    def normalize(self):
+        """
+        Nomalizes the qubit's quantum state
+        """
+        self.state = self.state / np.sqrt(np.dot(np.conjugate(self.state), self.state))
+
+    def get_probs(self):
+        """
+        Gives the probabilities of getting 0 or 1 when measuring the qubit
+        """
+        prob0 = (np.conjugate(self.state[0]) * self.state[0]).real
+        prob1 = (np.conjugate(self.state[1]) * self.state[1]).real
+        return prob0, prob1
+    
+
+class Agent:
+    """
+    Class representing the players in a quantum cryptography simulation (e.g. Alice, Eve or Bob)
+    """
+    def __init__(self, base=0, phase=0, key=None, message=None):
+        self.base = base
+        self.phase = phase
+
+    def project(self, qubit):
+        """
+        Projects the given qubit unto the agent's measuring basis
+        """
+        newqubit = qubit
+        base_delta = self.base - newqubit.base
+
+        # Projection oprator
+        proj_mtrx = np.array([
+            [np.cos(base_delta),                                -1 * np.exp(-1j * self.phase) * np.sin(base_delta)],
+            [np.sin(base_delta) * np.exp(-1j * self.phase),     np.cos(base_delta)                                ]
+        ])
+        
+        newqubit.state = np.dot(proj_mtrx, newqubit.state)
+        newqubit.normalize()
```

Comment:
I would need to refresh my memory on when should normalization take place.. should the qubit be normalized each time it gets projected? Is normalization affected by projection?

Reply:
The normalization can remove normalization, which is why the state has to be renormalized. You can read more about it [here](https://inst.eecs.berkeley.edu/~cs191/fa14/lectures/lecture89.pdf).